blog/content/post/2021-04-18-optiplex_part2.md

```diff
@@ -0,0 +1,249 @@
+along with the PS/2 keyboard and mouse. However the serial port stopped working
+whenever the machine lost the AC power. Noticing this small issue took me many
+days and on the first glance shouldn't prevent the board from booting, right??
+Well, the answer is: NO!
+
+**First of all, why it took me so many days to figure it out?**
+
+No matter how dumb it may sound like, but my builds were always with the
+onboard serial port enabled for debugging. Because why should I ever get rid of
+my main debugging method?? Of course there are other methods like USB debug
+dongle, but what for I have already a serial port... What is more I have
+verified it is working (at the beginning I had no idea it stops working after
+AC loss).
+
+**Secondly, I have interpreted the symptomes incorrectly**
+
+After the AC loss, the first thing I have noticed is that the led on the power
+button is not shining white. For clarification SCH5545 has a two-color LED
+driver that allows to change its colors based on the register initialization.
+Apparently the OptiPlex 9010 power button is equipped with an orange and white
+led. When the white led shines, it means the BIOS has started fetching and
+executing the code (which is good). However, if the registers are not
+configured correctly the power button was shining in orange color giving some
+error signal (the led was shining in intervals). My first conclusion was that
+the machine is not even starting after AC loss, because I had the code written
+for the leds to set them white and executed very early.
+
+Another symptome was the fan which started to run on full speed after a few
+seconds after I pressed power button. I thought initially, no big deal, if the
+machine is not starting it does not probably retrieve temperature from CPU or
+something.
+
+Summing up all those issues together I was almost certain it has something to
+do with the Super I/O and it prevents the board from booting. Of course the
+coreboot had not implemented all functionalities yet, but the bug was so severe
+that the board was not booting. I have looked into the datasheet of SCH5627 and
+noticed there is an Embedded Memory Interface (EMI) there which implements
+access to the Envrionmental Controller internal memory. "That must be it" I
+thought...
+
+![That chip...](/img/sch5545_yelling.jpg)
+
+<center>__*It is everything your fault!*__</center>
+
+Typically Super I/O chips are not that complex in design and operation, however
+those from SMSC stand out very much. There is no standardized access to
+hardware monitoring registers like temperature, fan tachometers and PWM for fan
+control (unlike the chips from ITE for example). So basically the manufacturer
+left the door wide open to developers and let them design the Environmental
+Controller firmware to perform the hardware monitoring and implement any
+communication protocol they want through the EMI. That doesn't sound good for
+me since I will have to go through reverse engineering path...
```

Comment:
@miczyg1 I would add more arguments that are aligned with EU Directive 2009/24/EC. We should explain why you had no other means of solving this problem? And what was your real target?

Also please be clear that you did that because of interoperability issues with coreboot and reverse engineering was done in your personal unpaid time for the benefit of community.

Reply:
Understood. Thanks for reference o the Directive.